lib/sdk/tabs/tab-firefox.js

```diff
@@ -3,309 +3,328 @@
-    if (!this._tab)
+  unpin: function() {
+    if (isDestroyed(this))
       return;
-    activateTab(this._tab);
-  }),
-  /**
-   * Close the tab
-   */
-  close: function close(callback) {
-    // Bug 699450: the tab may already have been detached
-    if (!this._tab || !this._tab.parentNode) {
+
+    unpin(viewsFor.get(this));
+  },
+
+  close: function(callback) {
+    let tabElement = viewsFor.get(this);
+
+    if (isDestroyed(this) || !tabElement || !tabElement.parentNode) {
       if (callback)
         callback();
       return;
     }
-    if (callback) {
-      if (this.window.tabs.activeTab && (this.window.tabs.activeTab.id == this.id))
-        observer.once('select', callback);
-      else
-        this.once(EVENTS.close.name, callback);
+
+    this.once('close', () => {
+      this.destroy();
+      if (callback)
+        callback();
+    });
+
+    closeTab(tabElement);
+  },
+
+  reload: function() {
+    if (isDestroyed(this))
+      return;
+
+    reload(viewsFor.get(this));
+  },
+
+  activate: defer(function() {
+    if (isDestroyed(this))
+      return;
+
+    activateTab(viewsFor.get(this));
+  }),
+
+  getThumbnail: function() {
+    if (isDestroyed(this))
+      return BLANK;
+
+    // TODO: This is unimplemented in e10s: bug 1148601
+    if (browser(this).isRemoteBrowser) {
+      console.error('This method is not supported with E10S');
+      return BLANK;
     }
-    this._window.gBrowser.removeTab(this._tab);
+    return getThumbnailURIForWindow(browser(this).contentWindow);
+  },
+
+  attach: function(options) {
+    if (isDestroyed(this))
+      return;
+
+    // BUG 792946 https://bugzilla.mozilla.org/show_bug.cgi?id=792946
+    // TODO: fix this circular dependency
+    let { Worker } = require('./worker');
+    return Worker(options, browser(this).contentWindow);
   },
-  /**
-   * Reload the tab
-   */
-  reload: function reload() {
-    if (!this._tab)
+
+  destroy: function() {
+    if (isDestroyed(this))
       return;
-    this._window.gBrowser.reloadTab(this._tab);
+
+    destroyed.set(this, true);
   }
 });
+exports.Tab = Tab;
 
-function getChromeTab(tab) {
-  return getOwnerWindow(viewNS(tab).tab);
-}
-
-// Implement `viewFor` polymorphic function for the Tab
-// instances.
-const getTabView = tab => viewNS(tab).tab;
+viewFor.define(Tab, tab => viewsFor.get(tab));
 
-function Tab(options, existingOnly) {
-  let chromeTab = options.tab;
-  for (let tab of TABS) {
-    if (chromeTab == tab._tab)
-      return tab._public;
+// Returns the high-level window for this DOM window if the windows module has
+// ever been loaded otherwise returns null
+function maybeWindowFor(domWindow) {
+  try {
+    return modelFor(domWindow);
   }
-  // If called asked to return only existing wrapper,
-  // we should return null here as no matching Tab object has been found
-  if (existingOnly)
+  catch (e) {
     return null;
+  }
+}
+
+function tabEmit(tab, event, ...args) {
+  const allTabs = require('../tabs');
```

Comment:
So this isn't what I'd consider to be a circular dependency, requiring tabs does automatically require tab-firefox but that doesn't automatically require tabs. Can you explain to me why this delayed require of tabs is a problem because I really don't get it?